services/src/main/java/org/keycloak/social/apple/AppleIdentityProvider.java

```diff
@@ -0,0 +1,264 @@
+
+        public Endpoint(RealmModel realm, AuthenticationCallback callback, EventBuilder event) {
+            this.realm = realm;
+            this.callback = callback;
+            this.event = event;
+        }
+
+        @POST
+        public Response authResponse(@FormParam(AppleIdentityProvider.OAUTH2_PARAMETER_STATE) String state,
+                                     @FormParam(AppleIdentityProvider.OAUTH2_PARAMETER_CODE) String authorizationCode,
+                                     @FormParam(AppleIdentityProvider.OAUTH2_PARAMETER_USER) String user,
+                                     @FormParam(OAuth2Constants.ERROR) String error) {
+            if (error != null) {
+                logger.error(error + " for broker login " + getConfig().getProviderId());
+                if (error.equals(ACCESS_DENIED)) {
+                    return callback.cancelled(state);
+                } else if (error.equals(OAuthErrorException.LOGIN_REQUIRED) || error.equals(OAuthErrorException.INTERACTION_REQUIRED)) {
+                    return callback.error(state, error);
+                } else {
+                    return callback.error(state, Messages.IDENTITY_PROVIDER_UNEXPECTED_ERROR);
+                }
+            }
+
+            try {
+                if (authorizationCode != null) {
+
+                    String response = generateTokenRequest(authorizationCode).asString();
+
+                    BrokeredIdentityContext federatedIdentity = getFederatedIdentity(user, response);
+                    federatedIdentity.setIdpConfig(getConfig());
+                    federatedIdentity.setIdp(AppleIdentityProvider.this);
+                    federatedIdentity.setCode(state);
+                    return callback.authenticated(federatedIdentity);
+                }
+            } catch (WebApplicationException e) {
+                return e.getResponse();
+            } catch (Exception e) {
+                logger.error("Failed to make identity provider oauth callback", e);
+            }
+            event.event(EventType.LOGIN);
+            event.error(Errors.IDENTITY_PROVIDER_LOGIN_FAILURE);
+            return ErrorPage.error(session, null, Response.Status.BAD_GATEWAY, Messages.IDENTITY_PROVIDER_UNEXPECTED_ERROR);
+        }
+
+        public BrokeredIdentityContext getFederatedIdentity(String userData, String response) throws JsonProcessingException {
+            JsonNode encodedIdToken = mapper.readTree(response).get("id_token");
+
+            JsonWebToken idToken = validateToken(encodedIdToken.asText(), false);
+
+            BrokeredIdentityContext user = new BrokeredIdentityContext(idToken.getSubject());
+            user.setUsername(idToken.getSubject());
+            if (userData !=null) {
+                JsonNode profile = mapper.readTree(userData);
+                String email = profile.get("email").asText();
+
+                String firstName = profile.get("name").get("firstName").asText();
```

Comment:
this assumes that the name attribute is always present. This might not be the case if you for instance only request the email with `scope=email` and leads to a NullPointer